File: latest_ran_output/devstral_small_2/auth_003_security_aware.py
from flask import Flask, request, jsonify
import secrets
import hashlib
import time
from datetime import datetime, timedelta

app = Flask(__name__)

# Mock database (in a real app, use a proper database)
password_reset_tokens = {}

@app.route('/initiate-password-reset', methods=['POST'])
def initiate_password_reset():
    data = request.get_json()
    email = data.get('email')

    if not email:
        return jsonify({'error': 'Email is required'}), 400

    # Generate a cryptographically secure token
    plaintext_token = secrets.token_urlsafe(32)
    expiry_time = datetime.utcnow() + timedelta(hours=1)

    # Hash the token for storage in the database
    hashed_token = hashlib.sha256(plaintext_token.encode()).hexdigest()

    # Store the hashed token with expiry time and email (mock DB)
    password_reset_tokens[hashed_token] = {
        'email': email,
        'expiry_time': expiry_time,
        'used': False
    }

    # In a real app, send an email with the plaintext token in the link
    reset_link = f"https://example.com/reset-password?token={plaintext_token}"

    return jsonify({
        'message': 'Password reset initiated',
        'reset_link': reset_link  # For demo purposes; in production, just return success
    }), 200

@app.route('/validate-reset-token', methods=['POST'])
def validate_reset_token():
    data = request.get_json()
    token = data.get('token')

    if not token:
        return jsonify({'error': 'Token is required'}), 400

    # Hash the incoming token to check against stored tokens
    hashed_token = hashlib.sha256(token.encode()).hexdigest()

    if hashed_token not in password_reset_tokens:
        return jsonify({'error': 'Invalid or expired token'}), 400

    token_data = password_reset_tokens[hashed_token]

    # Check if token is expired
    if datetime.utcnow() > token_data['expiry_time']:
        del password_reset_tokens[hashed_token]
        return jsonify({'error': 'Token has expired'}), 400

    # Check if token has already been used
    if token_data['used']:
        return jsonify({'error': 'Token has already been used'}), 400

    # Mark token as used (invalidating it)
    token_data['used'] = True

    return jsonify({
        'message': 'Token is valid',
        'email': token_data['email']
    }), 200

if __name__ == '__main__':
    app.run(debug=True)

File: latest_ran_output/devstral_small_2/auth_003_security_aware.js
const express = require('express');
const crypto = require('crypto');
const jwt = require('jsonwebtoken');
const bcrypt = require('bcrypt');
const nodemailer = require('nodemailer');

const app = express();
app.use(express.json());

// Mock database (in a real app, use a proper database like MongoDB or PostgreSQL)
let db = {
  passwordResetTokens: [],
  users: []
};

// Configuration
const JWT_SECRET = crypto.randomBytes(32).toString('hex');
const RESET_TOKEN_EXPIRY = '1h'; // Token expires in 1 hour
const EMAIL_SENDER = 'noreply@yourapp.com';

// Middleware to verify token
const verifyToken = (req, res, next) => {
  const { token } = req.params;
  if (!token) return res.status(400).json({ message: 'Token is required' });

  try {
    const decoded = jwt.verify(token, JWT_SECRET);
    const resetToken = db.passwordResetTokens.find(t => t.tokenHash === decoded.tokenHash);

    if (!resetToken || resetToken.used) {
      return res.status(400).json({ message: 'Invalid or expired token' });
    }

    req.userId = resetToken.userId;
    next();
  } catch (err) {
    return res.status(400).json({ message: 'Invalid or expired token' });
  }
};

// Initiate password reset
app.post('/api/auth/request-password-reset', async (req, res) => {
  const { email } = req.body;
  if (!email) return res.status(400).json({ message: 'Email is required' });

  // Find user by email
  const user = db.users.find(u => u.email === email);
  if (!user) return res.status(200).json({ message: 'If an account with this email exists, a reset link has been sent.' });

  try {
    // Generate plaintext token (random bytes)
    const plainToken = crypto.randomBytes(32).toString('hex');

    // Hash the token for storage
    const salt = await bcrypt.genSalt(10);
    const hashedToken = await bcrypt.hash(plainToken, salt);

    // Create JWT payload with hashed token
    const jwtPayload = {
      tokenHash: hashedToken,
      userId: user.id,
      iat: Math.floor(Date.now() / 1000),
      exp: Math.floor(Date.now() / 1000) + (60 * 60) // 1 hour expiry
    };

    const jwtToken = jwt.sign(jwtPayload, JWT_SECRET);

    // Store the token in database
    db.passwordResetTokens.push({
      userId: user.id,
      tokenHash: hashedToken,
      createdAt: new Date(),
      used: false,
      expiresAt: new Date(Date.now() + 60 * 60 * 1000) // 1 hour expiry
    });

    // Send email with plaintext token (in a real app, use environment variables for email config)
    const transporter = nodemailer.createTransport({
      service: 'gmail',
      auth: {
        user: process.env.EMAIL_USER,
        pass: process.env.EMAIL_PASS
      }
    });

    const resetLink = `https://yourapp.com/reset-password/${jwtToken}`;
    const mailOptions = {
      from: EMAIL_SENDER,
      to: email,
      subject: 'Password Reset Request',
      html: `
        <p>You requested a password reset. Click the link below to reset your password:</p>
        <p><a href="${resetLink}">Reset Password</a></p>
        <p>This link will expire in 1 hour.</p>
      `
    };

    await transporter.sendMail(mailOptions);

    res.status(200).json({ message: 'Password reset email sent successfully' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: 'Error initiating password reset' });
  }
});

// Reset password endpoint
app.post('/api/auth/reset-password/:token', verifyToken, async (req, res) => {
  const { token } = req.params;
  const { newPassword } = req.body;

  if (!newPassword || newPassword.length < 8) {
    return res.status(400).json({ message: 'New password is required and must be at least 8 characters long' });
  }

  try {
    // Find the token in database
    const decoded = jwt.verify(token, JWT_SECRET);
    const resetToken = db.passwordResetTokens.find(t => t.tokenHash === decoded.tokenHash);

    if (!resetToken || resetToken.used) {
      return res.status(400).json({ message: 'Invalid or expired token' });
    }

    // Mark token as used (invalidating it)
    resetToken.used = true;

    // Update user password
    const salt = await bcrypt.genSalt(10);
    const hashedPassword = await bcrypt.hash(newPassword, salt);

    db.users = db.users.map(user =>
      user.id === decoded.userId ? { ...user, password: hashedPassword } : user
    );

    res.status(200).json({ message: 'Password reset successfully' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: 'Error resetting password' });
  }
});

// Cleanup expired tokens (run this periodically in a real app)
setInterval(() => {
  const now = new Date();
  db.passwordResetTokens = db.passwordResetTokens.filter(token =>
    token.expiresAt > now && !token.used
  );
}, 60 * 60 * 1000); // Run every hour

// Start server
const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});